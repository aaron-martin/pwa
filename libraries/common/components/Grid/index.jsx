/**
 * Copyright (c) 2017-present, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { objectWithoutProps } from '../../helpers/data';
import GridItem from './components/Item';
import styles, { wrap } from './style';

/**
 * The grid component.
 */
class Grid extends Component {
  static Item = GridItem;

  static propTypes = {
    className: PropTypes.string,
    component: PropTypes.string,
    wrap: PropTypes.bool,
  };

  static defaultProps = {
    className: '',
    component: 'ul',
    wrap: false,
  };

  /**
   * Composes the props.
   * @returns {Object} The composed props.
   */
  getProps() {
    let className = `${this.props.className} ${styles}`;

    if (this.props.wrap) {
      className += ` ${wrap(this.props.wrap)}`;
    }

    const props = {
      ...this.props,
      className,
    };

    return objectWithoutProps(props, [
      'wrap',
      'component',
    ]);
  }

  /**
   * Renders the component.
   * @returns {JSX}
   */
  render() {
    return React.createElement(this.props.component, this.getProps());
  }
}

export default Grid;
