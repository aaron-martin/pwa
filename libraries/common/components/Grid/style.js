/**
 * Copyright (c) 2017, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { css } from 'glamor';

/**
 * Creates a class name for the flex-wrap property
 * @param {boolean} [value=false] TRUE for 'wrap', FALSE for 'nowrap'
 * @return {string} The class name
 */
export const wrap = (value = false) => css({
  flexWrap: value ? 'wrap' : 'nowrap',
}).toString();

export default css({
  display: 'flex',
  minWidth: '100%',
}).toString();
