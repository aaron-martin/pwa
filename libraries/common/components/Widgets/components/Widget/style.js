/**
 * Copyright (c) 2017-present, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { css } from 'glamor';

const content = css({
  position: 'absolute',
  top: 0,
  left: 0,
  right: 0,
  bottom: 0,
  overflow: 'hidden',
}).toString();

/**
 * Calculates the height of a widget.
 * @param {number} base The cell size of the widget grid.
 * @param {number} rows The number of rows to use.
 * @returns {Function}
 */
const height = (base, rows) => css({
  display: 'block',
  position: 'absolute',
  height: `${Math.ceil(base * rows)}px`,
}).toString();

/**
 * Calculates the left offset based on the given number of columns as a faction of 12.
 * @param {number} offset The number of columns to offset by.
 * @returns {Function}
 */
const left = (offset) => {
  if (offset === 0) {
    return '';
  }

  return css({
    left: `${((offset * 100) / 12).toFixed(5)}%`,
  }).toString();
};

/**
 * Calculates the top offset based on the given cell size and row number.
 * @param {number} base The cell size of the widget grid.
 * @param {number} offset The number of columns to offset by.
 * @returns {Function}
 */
const top = (base, offset) => {
  if (offset === 0) {
    return '';
  }

  return css({
    top: `${Math.ceil(offset * base)}px`,
  }).toString();
};

/**
 * Calculates the width based on the given number of columns as a faction of 12.
 * @param {number} columns The number of columns to use.
 * @returns {Function}
 */
const width = (columns) => {
  if (columns === 12) {
    return css({
      width: '100%',
    }).toString();
  }

  return css({
    width: `${((columns * 100) / 12).toFixed(5)}%`,
  }).toString();
};

export default {
  content,
  height,
  left,
  top,
  width,
};
