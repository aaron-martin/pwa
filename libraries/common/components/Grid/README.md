# Grid
---
The Grid component is a **Higher Order Component** that acts as a `flexbox` wrapper. The child elements will be rendered as a grid.

## Getting Started

```markup
import { Grid } from 'Library/components';

<Grid>
  <MyChildComponent />
  <MyChildComponent />
  <MyOtherChildComponent />
</Grid>
```

## Props

### component

_Default_: `ul`  
_Type_: `string`  

The HTML tag that will be used to render the component.

###### Usage:

```
<Grid component="article" />
```

### wrap

_Default_: `false`   
_Type_: `boolean`  

Applies `flex-wrap: [wrap | nowrap]` to the component.

###### Usage:

```
<Grid wrap />
```

### className

_Default_: `''`  
_Type_: `string`  

Additional classes that can be appended to the component's `className` attribute.

###### Usage:
```
<Grid className="class1 class2" />
```
---
