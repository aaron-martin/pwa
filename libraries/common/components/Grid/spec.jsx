/**
 * Copyright (c) 2017-present, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import { shallow } from 'enzyme';
import Grid from './index';

describe('<Grid />', () => {
  it('should render without any further props', () => {
    const wrapper = shallow(<Grid />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should be able to render a custom tag', () => {
    const wrapper = shallow(<Grid component="article" />);

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.type()).toEqual('article');
  });

  it('should add custom classes on demand', () => {
    const wrapper = shallow(<Grid className="custom-class-name" />);

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.hasClass('custom-class-name')).toEqual(true);
  });
});
