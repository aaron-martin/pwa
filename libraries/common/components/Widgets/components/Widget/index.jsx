/**
 * Copyright (c) 2017, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import PropTypes from 'prop-types';
import Grid from '../../../Grid';
import styles from './style';

/**
 * The widget component.
 * @param {Object} props The component props.
 * @returns {JSX}
 */
const Widget = (props) => {
  const { col, row, height, settings, width } = props.config;
  const { cellSize } = props;

  if (!props.component) {
    return null;
  }

  let className = '';
  if (cellSize) {
    className = [
      styles.height(cellSize, height, width),
      styles.width(width),
      styles.top(cellSize, row, height),
      styles.left(col),
    ].join(' ');
  }

  return (
    <Grid.Item
      className={className}
      component="div"
    >
      <div className={styles.content}>
        {React.createElement(props.component, {
          settings,
          ratio: [width, height],
        })}
      </div>
    </Grid.Item>
  );
};

Widget.propTypes = {
  config: PropTypes.shape().isRequired,
  cellSize: PropTypes.number,
  component: PropTypes.func,
};

Widget.defaultProps = {
  cellSize: null,
  component: null,
};

export default Widget;
