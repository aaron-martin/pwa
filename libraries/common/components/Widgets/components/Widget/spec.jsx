/**
 * Copyright (c) 2017, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import { mount } from 'enzyme';
import Grid from '../../../Grid';
import Widget from './index';

/**
 * A dummy component.
 * @returns {JSX}
 */
const MyComponent = () => (
  <div />
);

const widgets = {
  'core-widgets/image': MyComponent,
};

describe('<Widget />', () => {
  it('should render an image widget', () => {
    const config = {
      type: 'core-widgets/image',
      col: 1,
      row: 1,
      width: 12,
      height: 6,
    };

    const wrapper = mount(
      <Widget config={config} component={widgets[config.type]} />
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(Grid.Item).exists()).toBe(true);
  });

  it('should not render when the `type` prop is invalid', () => {
    const config = {
      type: 'some_widget', // Invalid
      col: 1,
      row: 1,
      width: 12,
      height: 6,
    };

    const wrapper = mount(
      <Widget config={config} component={widgets[config.type]} />
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(Grid.Item).exists()).toBe(false);
  });
});
