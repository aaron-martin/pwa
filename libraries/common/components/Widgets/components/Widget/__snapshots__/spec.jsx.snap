// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Widget /> should not render when the \`type\` prop is invalid 1`] = `
<Widget
  cellSize={null}
  component={null}
  config={
    Object {
      "col": 1,
      "height": 6,
      "row": 1,
      "type": "some_widget",
      "width": 12,
    }
  }
/>
`;

exports[`<Widget /> should render an image widget 1`] = `
<Widget
  cellSize={null}
  component={[Function]}
  config={
    Object {
      "col": 1,
      "height": 6,
      "row": 1,
      "type": "core-widgets/image",
      "width": 12,
    }
  }
>
  <GridItem
    className=""
    component="div"
    grow={0}
    shrink={1}
  >
    <div
      className=""
    >
      <div
        className="css-4hpmhj"
      >
        <MyComponent
          ratio={
            Array [
              12,
              6,
            ]
          }
          settings={undefined}
        >
          <div />
        </MyComponent>
      </div>
    </div>
  </GridItem>
</Widget>
`;
